techlibs/intel_alm/common/quartus_rename.v

```
@@ -2,11 +2,19 @@
 `define LCELL cyclonev_lcell_comb
 `define MAC cyclonev_mac
 `define MLAB cyclonev_mlab_cell
+`define RAM_BLOCK cyclonev_ram_block
```

call it M10K_BLOCK perhaps?

(there are Intel FPGAs with several different sizes of memory blocks, could be good to establish the convention here to specify size in case we end up adding support for them)

---

I fully intend to add native instantiation for Cyclone 10 GX, which uses M20K blocks instead. The problem then is what to define M10K_BLOCK on a chip that doesn't have it, and vice versa.

I think there *might* have been a family that mixed M144Ks with M20Ks, but I can't remember it. If not, then I think it's reasonably safe to use RAM_BLOCK to refer to BRAM.

---

Just don't define it and disable the substitution or have it `$error`

Converting a wrong type of primitive is a silent failure of the kind that we would really like to avoid

---

According to my notes, the only mixed-size families that involve ALMs are stratix iii and stratix iv (aka arria ii), both of them involving M9K+M144K; still, this does sound like something we could support one day

---

> Just don't define it and disable the substitution

I'm pretty sure an unexpanded substitution is a syntax error which would fail whenever techmap was called.

`$error` is probably okay though.

---

It's actually a "too hard" error — it will fail reading the whole techmap file whether the invalid substitution is in fact used or not, so it needs another `ifdef